syntax = "proto2";

package request;

message Operation {

  enum OperationType {
    READ = 0;
    WRITE = 1;
  }

  required OperationType type = 1;  // Type of operation: READ or WRITE
  required string key = 2;    // The data item on which the operation is being performed
  optional string value = 3;        // Value to be written if it's a WRITE operation (not needed for READ)
}

message Transaction {
  required int32 order = 1;
  required string id = 2;
  repeated Operation operations = 3;  // A list of operations that are part of the transaction
  optional int32 client_id = 4; // Client ID requesting the transaction
}

message Request {

  enum RequestRecipient {
    BATCHER = 0;
    PARTIAL = 1;
    MERGER = 3;
    PING = 4;
  }

  optional int32 client_id = 1;     // Client ID making the request
  optional int32 server_id = 2;     // Server ID making the request
  repeated Transaction transaction = 3;  // The transactions being requested
  required RequestRecipient recipient = 4; // The recipient of the request

}